### PR TITLE
termio: preserve UTF-8 in desktop notification truncation

### LINKED ISSUE #13795: OSC 9 notification produces [Invalid UTF-8] with multibyte UTF‑8 characters in msg
### Discussed in https://github.com/ghostty-org/ghostty/discussions/13790

<div type='discussions-op-text'>

<sup>Originally posted by **dolzenko** August 13, 2026</sup>
### Issue Description

Originally when using ghostty with codex cli and conversation contains Russian chars I was getting 
image
on task completion.

### Expected Behavior

Proper notification message

### Actual Behavior

Notification doesn't appear

### Reproduction Steps

1. Open Ghostty
2. Run `bash -c 'p=$(printf "a%.0s" {1..254}); printf "\033]9;%sЯ\007" "$p"'`

### Ghostty Logs

_No response_

### Ghostty Version

```text
Ghostty 1.3.1-arch2

Version
  - version: 1.3.1-arch2
  - channel: tip
Build Config
  - Zig version   : 0.15.2
  - build mode    : .ReleaseFast
  - app runtime   : .gtk
  - font engine   : .fontconfig_freetype
  - renderer      : renderer.generic.Renderer(renderer.OpenGL)
  - libxev        : io_uring
  - kernel version: 7.1.6-arch1-1
  - desktop env   : other
  - GTK version   :
    build         : 4.22.2
    runtime       : 4.22.4
  - libadwaita    : enabled
    build         : 1.9.0
    runtime       : 1.9.3
  - libX11        : enabled
  - libwayland    : enabled
```

### OS Version Information

arch

### (Linux only) Display Server

Wayland

### (Linux only) Desktop Environment/Window Manager

_No response_

### Minimal Ghostty Configuration

```ini
# No configuration required.
```

### Additional Relevant Configuration

_No response_

### AI Usage Disclosure

I used OpenAI Codex to help investigate the root cause, implement and test a candidate fix, and edit this report. I reviewed the report and code changes.

### I acknowledge that:

- [X] I have reviewed the FAQ and confirm that my issue is NOT among them.
- [X] I have searched the Ghostty documentation and confirm there doesn't exist a configuration option that solves my issue.
- [X] I have searched the Ghostty repository (both open and closed Discussions and Issues) and confirm this is not a duplicate of an existing issue or discussion.
- [X] I have checked the "Preview" tab on all text fields to ensure that everything looks right, and have wrapped all configuration and code in code blocks with a group of three backticks (` ``` `) on separate lines.</div>

### PR DESCRIPTION
**Pull request by dolzenko**
## Summary

- Prevent desktop notification title and body truncation from producing invalid UTF-8.
- Truncate fixed-size buffers to the longest valid UTF-8 prefix.
- Add regression tests for multibyte characters at the buffer boundary.

Fixes #13795

## Testing

- Confirmed the original reproducer produces `[Invalid UTF-8]` with the installed Ghostty.
- Confirmed the patched Ghostty displays a valid, truncated notification.
- Added tests covering both notification title and body truncation.

## AI Usage Disclosure

I used OpenAI Codex to investigate the root cause, implement the fix and regression tests, run validation, and help prepare the issue and PR descriptions. I reviewed and understand the submitted change.

### CODE REVIEW

File: src/termio/stream_handler.zig [hunk 1]
```diff
@@ -16,6 +16,17 @@ const posix = std.posix;
 
 const log = std.log.scoped(.io_handler);
 
+/// Copy the longest valid UTF-8 prefix that fits in a sentinel-terminated
+/// destination. Desktop notification strings are passed to platform APIs that
+/// require valid UTF-8, so truncation must not split a multibyte codepoint.
+fn copyUtf8Z(comptime capacity: usize, dst: *[capacity:0]u8, src: []const u8) void {
+    var len = @min(src.len, capacity);
+    while (len > 0 and !std.unicode.utf8ValidateSlice(src[0..len])) : (len -= 1) {}
```
Comment (bo2themax) [inline src/termio/stream_handler.zig]: This seems O(n²) when capacity is 0

Comment (dolzenko) [reply src/termio/stream_handler.zig]: Good catch.  I replaced that in `ae6d97e` with a backward scan over continuation bytes at the truncation boundary. It takes at most three steps for valid UTF-8 and O(capacity) for malformed continuation runs, without rescanning prefixes. I also added tests covering 1–4-byte sequences and a complete codepoint at the boundary.

**Review by mitchellh — CHANGES_REQUESTED** [review]
I think this function perhaps belongs elsewhere (not sure atm) but this functionality makes sense.

**Comment by jcollie** [comment]
> I think this function perhaps belongs elsewhere (not sure atm) but this functionality makes sense.

`src/lib/strings.zig`?

File: src/termio/stream_handler.zig [hunk 2]
```diff
@@ -1348,13 +1369,16 @@ pub const StreamHandler = struct {
...
-        const title_len = @min(title.len, message.desktop_notification.title.len);
-        @memcpy(message.desktop_notification.title[0..title_len], title[0..title_len]);
-        message.desktop_notification.title[title_len] = 0;
-
-        const body_len = @min(body.len, message.desktop_notification.body.len);
-        @memcpy(message.desktop_notification.body[0..body_len], body[0..body_len]);
-        message.desktop_notification.body[body_len] = 0;
+        copyUtf8Z(
+            message.desktop_notification.title.len,
+            &message.desktop_notification.title,
```
Comment (jcollie) [inline src/termio/stream_handler.zig]: `&` shouldn't be necessary to pass a `[] const u8`

**Review by jcollie — APPROVED** [review]

**Comment by dolzenko** [comment]
> I think this function perhaps belongs elsewhere (not sure atm) but this functionality makes sense.

Addressed in 53c6fdbe: moved the fixed-buffer representation and UTF-8 truncation into `Message.DesktopNotification.init`, so `stream_handler` now only constructs and sends the message.

File: src/termio/stream_handler.zig (hunk 2 above)
Comment (dolzenko) [reply src/termio/stream_handler.zig]: > `&` shouldn't be necessary to pass a `[] const u8`

But 2nd arg is not const its the mutable destination pointer?

**Review by jcollie — APPROVED** [review]
Looks a lot cleaner.
